### PR TITLE
 ⭐️ New: Add rule `no-reserved-component-names`

### PR DESCRIPTION
**Pull request by shadskii**
This rule addresses #486 and prevents name collisions with vue component and standard html elements. Both PascalCase names and kebab-case names of standard html elements are invalid. For example, `'div'` and `'Div'` are both treated as reserved names.

### CODE REVIEW

File: lib/rules/no-reserved-component-names.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'disallow the use of reserved names in component definitions',
+      category: 'essential',
```
Comment (ota-meshi) [inline lib/rules/no-reserved-component-names.js]: Please set undefined.

```suggestion
      category: undefined, // 'essential'
```

Categories are set at the timing of the major version up.

File: lib/rules/no-reserved-component-names.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  'color-profile',
+  'font-face',
+  'font-face-src',
+  'font-face-uri',
+  'font-face-format',
+  'font-face-name',
+  'missing-glyph'
+]
+
+const isLowercase = (word) => (/[a-z]/.test(word))
```
Comment (ota-meshi) [inline lib/rules/no-reserved-component-names.js]: `/^[a-z]*$/.test(word)` ?

File: docs/rules/no-reserved-component-names.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+</script>
+```
+
+</eslint-code-block>
+
+## :books: Further reading
+
+- [List of html elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element)
+- [List of SVG elements](https://developer.mozilla.org/en-US/docs/Web/SVG/Element)
+- [Kebab case elements](https://stackoverflow.com/questions/22545621/do-custom-elements-require-a-dash-in-their-name/22545622#22545622)
```
Comment (ota-meshi) [inline docs/rules/no-reserved-component-names.md]: I think that the next link is the source, so can you add it as well?

https://w3c.github.io/webcomponents/spec/custom/#valid-custom-element-name

File: lib/rules/no-reserved-component-names.js (hunk 2 above)
Comment (shadskii) [reply lib/rules/no-reserved-component-names.js]: Yes, that is a more accurate regex!

File: lib/rules/no-reserved-component-names.js (hunk 1 above)
Comment (shadskii) [reply lib/rules/no-reserved-component-names.js]: My mistake, I have corrected this with your suggestion!

**Comment by shadskii** [comment]
I've added warnings for locally registered components as @michalsnik suggested in #486.
For example:
```javascript
export default {
  components: {
    'font-face': {},
  }
}
```
Thanks for the suggestion!

File: tests/lib/rules/no-reserved-component-names.js [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+  'textarea', 'Textarea',
+  'details', 'Details',
+  'dialog', 'Dialog',
+  'menu', 'Menu',
+  'menuitem', 'menuitem',
+  'summary', 'Summary',
+  'content', 'Content',
+  'element', 'Element',
+  'shadow', 'Shadow',
+  'template', 'Template',
```
Comment (Mouvedia) [inline tests/lib/rules/no-reserved-component-names.js]: [`<template>`](https://vuejs.org/v2/guide/list.html#v-for-on-a-lt-template-gt) should be allowed right?

cf https://html.spec.whatwg.org/multipage/scripting.html#the-template-element

**Comment by Mouvedia** [comment]
The `nobr`, `strike` and `acronym` tags are missing from [this list](https://github.com/vuejs/eslint-plugin-vue/pull/757/files#diff-05040e918e9890028624bfff94c813d4R18).

File: tests/lib/rules/no-reserved-component-names.js [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+  'slot', 'Slot',
```
Comment (Mouvedia) [inline tests/lib/rules/no-reserved-component-names.js]: Shouldn't this conflict with [slots](https://vuejs.org/v2/guide/components-slots.html)?

**Comment by shadskii** [comment]
> The `nobr`, `strike` and `acronym` tags are missing from [this list](https://github.com/vuejs/eslint-plugin-vue/pull/757/files#diff-05040e918e9890028624bfff94c813d4R18).

Originally, I chose to only check for non-obsolete elements. `nobr`, `strike` and `acronym` are all obsolete (see their MDN reference [`nobr`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nobr), [`strike`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/strike), and [`acronym`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/acronym) ). You bring up a good question: _Should we check for obsolete or deprecated tags?_ I'm inclined to believe the answer is _yes_. Browsers still render deprecated tags and this will result in a name collision if a vue component shares the same name. 

MDN lists 31 obsolete and deprecated elements (see this [list](https://developer.mozilla.org/en-US/docs/Web/HTML/Element#Obsolete_and_deprecated_elements)). I have added `deprecated-html-elements.json` with these elements and updated this rule to search for them. 

Thank you for reviewing this pull request and please let me know if you agree with the updates I've made!

File: tests/lib/rules/no-reserved-component-names.js (hunk 4 above)
Comment (shadskii) [reply tests/lib/rules/no-reserved-component-names.js]: Yes, I've added a test to validate this.

File: tests/lib/rules/no-reserved-component-names.js (hunk 5 above)
Comment (shadskii) [reply tests/lib/rules/no-reserved-component-names.js]: Yes, same for `slot`, I've added a test to validate.

**Comment by Mouvedia** [comment]
> I'm inclined to believe the answer is _yes_.

I agree as well, but I wanted to illustrate the need with elements that are obsolete but still heavily exploited nowadays.

> Thank you for reviewing this pull request and please let me know if you agree with the updates I've made!

These were merely comments. I didn't review your code nor do I have the time to.
Thanks _you_ for the PR.

**Pull request merged**